skymp5-server/cpp/server_guest_lib/AnimationSystem.cpp

```diff
@@ -3,18 +3,21 @@
 #include "AnimationData.h"
 #include "MathUtils.h"
 #include "MpActor.h"
+#include "WorldState.h"
 #include "libespm/espm.h"
 #include <unordered_map>
 
-AnimationSystem::AnimationSystem(bool isSweetpie, WorldState& worldState_)
-  : worldState(worldState_)
+void AnimationSystem::Init(WorldState* pWorldState)
 {
-  InitAnimationCallbacks(isSweetpie);
+  if (!pWorldState) {
+    spdlog::error("No worldState attached to animation system. Using default "
+                  "values for stamina forfeits");
+  }
+  hasSweetpie = pWorldState->HasEspmFile("SweetPie.esp");
```

Comment:
potential null pointer dereference here.

please use WorldState&, I don't think we need pointers here

Reply:
To store a reference in a class we have to provide it during object initialization. Here, I made `AnimationSystem` to be initialized on stack in `PartOne`, and we want  to set pointer to worldState somewhen later, when there is espm attached to it. This happens [here](https://github.com/skyrim-multiplayer/skymp/pull/1671/files#diff-c35c3fcadfee3ad4d1ada7e1172ffc7eac60c0e3c60c369ab71a886a025d37ceR303). At this moment we're unable to use reference, so this is why it is a pointer. As an alternative, I can make initialization of `AnimationSystem` to be on heap again and initialize it at right moment and then set map with stamina forfeits. So, what do I have to do?

Reply:
I think you can use this syntax to pass by reference `partOne->animationSystem.Init(partOne->worldState);`

It seems that it should work. Because under the hood reference = pointer.

Reply:
It will be a compiler-error in this case. Take a look at [this](https://stackoverflow.com/questions/10956139/uninitialized-reference-member)

Reply:
make it a constructor then

Reply:
I'll accept any variant where there is no code path that dereferences null pointer

you can add return after logs in your original code if you wish. or refactor as you wish.

but my point is that there shouldn't be `if (!x) { log() } x->y;`

Reply:
There actually isn't. [This](https://github.com/skyrim-multiplayer/skymp/pull/1671/files#diff-4481088266a71f94bd45bb76535e5f69628d98e8f7fb395635d14e399e86c23fR16) one was outdated. So should I leave it as is, or refactor to be a reference?

Reply:
Well, I have redone this a little bit. I think it looks better now. [click](https://github.com/skyrim-multiplayer/skymp/pull/1671/files#diff-4481088266a71f94bd45bb76535e5f69628d98e8f7fb395635d14e399e86c23fR12)